group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -0,0 +1,415 @@
+        if (maybeAssignPendingPartitions(newAssignedSet, newAssigningSet)) {
+            return new ConsumerGroupMember.Builder(member)
+                .setAssignedPartitions(newAssignedSet)
+                .setPartitionsPendingRevocation(Collections.emptyMap())
+                .setPartitionsPendingAssignment(newAssigningSet)
+                .setPreviousMemberEpoch(member.memberEpoch())
+                .setMemberEpoch(targetAssignmentEpoch)
+                .setNextMemberEpoch(targetAssignmentEpoch)
+                .build();
+        } else {
+            return member;
+        }
+    }
+
+    /**
+     * Makes a deep copy of an assignment map.
+     *
+     * @param map The Map to copy.
+     * @return The copy.
+     */
+    private Map<Uuid, Set<Integer>> deepCopy(Map<Uuid, Set<Integer>> map) {
+        Map<Uuid, Set<Integer>> copy = new HashMap<>();
+        map.forEach((topicId, partitions) -> {
+            copy.put(topicId, new HashSet<>(partitions));
+        });
+        return copy;
+    }
+
+    /**
+     * Tries to move partitions from the Assigning set to the Assigned set
+     * if they are no longer owned.
+     *
+     * @param newAssignedSet  The Assigned set.
+     * @param newAssigningSet The Assigning set.
+     * @return A boolean indicating if any partitions were moved.
+     */
+    private boolean maybeAssignPendingPartitions(
+        Map<Uuid, Set<Integer>> newAssignedSet,
+        Map<Uuid, Set<Integer>> newAssigningSet
+    ) {
+        boolean changed = false;
+
+        Iterator<Map.Entry<Uuid, Set<Integer>>> assigningSetIterator = newAssigningSet.entrySet().iterator();
+        while (assigningSetIterator.hasNext()) {
+            Map.Entry<Uuid, Set<Integer>> pair = assigningSetIterator.next();
+            Uuid topicId = pair.getKey();
+            Set<Integer> assigning = pair.getValue();
+            Iterator<Integer> assigningIterator = assigning.iterator();
+            while (assigningIterator.hasNext()) {
+                Integer partitionId = assigningIterator.next();
+                Integer partitionEpoch = currentPartitionEpoch.apply(topicId, partitionId);
+                if (partitionEpoch == -1) {
+                    assigningIterator.remove();
+                    put(newAssignedSet, topicId, partitionId);
+                    changed = true;
+                }
+            }
+            if (assigning.isEmpty()) {
+                assigningSetIterator.remove();
+            }
+        }
+
+        return changed;
+    }
+
+    /**
+     * Checks whether the owned topic partitions passed by the member to the state
+     * machine via the ConsumerGroupHeartbeat request corresponds to the Assigned
+     * set.
+     *
+     * @param ownedTopicPartitions The topic partitions owned by the remove client.
+     * @return A boolean indicating if the owned partitions matches the Assigned set.
+     */
+    private boolean hasRevokedAllPartitions(
```

Comment:
From the name of hasRevokedAllPartitions seems it checks whether a client removed all the partitions but it actually checks whether a client owns the correct set of assigned partitions. Do I misunderstand too much?

Reply:
Yeah -- if we are checking the correct state after revoking, this name is a bit confusing

Reply:
Also would we expect this to match assigned set after we assign? We call this before we assign.

Reply:
@CalvinConfluent Your understanding is correct.
@jolshan No. This is the condition to transition to Assigning state. We basically want to consumer to ack'ed that it has revoked partitions before we assign new partitions.

Let me see if I can rename it. I called it `hasRevokedAllPartitions` in the first place because this is actually what we want from the member.

Reply:
The owned partitions are provided in the ConsumerGroupHeartbeat request. They correspond to what the member currently has assigned to it (from his point of view). If this corresponds to the assigned partitions, it means that the member is done with revoking the partitions.

The assigned partition set is the set of partitions currently assigned to the member. Those are the partitions being processed/fetched by the member.

I renamed `assigning set` to `partitions pending assignment` to make thing clearer. I hope it helps.

The partitions pending assignment are the partitions that the member will eventually receive when it has confirmed that it has revoked the partitions that it no longer owns. The pending partitions will be assigned to the member if they are not owned by any other members or when they are freed up.

Reply:
Ack. I think in my mind, the distinction was when we reached the final target state. But really stage 1 is getting to the state where we remove all unnecessary partitions and then bump the epoch. Then we proceed to stage 2 where we assign the ones that are missing. I think I was viewing "assigned" as the final state, but really this is just the final state of stage 1. (and target will be the final state of stage 2)